### DISCUSSION TITLE
Added: C# Language Design Notes for Apr 18, 2017

### DISCUSSION THREAD

**Discussion by MadsTorgersen** [dotnet/csharplang#651] [General]
[C# Language Design Notes for Apr 18, 2017](https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-04-18.md)

1. Default implementations for event accessors in interfaces
2. Reabstraction in a class of default-implemented member
3. `sealed override` with default implementations
4. Use of `sealed` keyword for non-virtual interface members
5. Implementing inaccessible interface members
6. Implicitly implementing non-public interface members
7. Not quite implementing a member
8. asynchronous `Main`

Please discuss.

**Comment by svick** [comment]
> What about other awaitable types? Should they be allowed? This would be easy enough to implement; the concern is whether it is reasonable to expect their `GetResult()` method to block? After all, the compiler has not previously relied on this in its use of `GetResult()` on awaitable types, so one would assume that no particular effort has been put into ensuring it in the implementation of those types.
>
> If we don't allow other awaitables in `Main` directly, folks can easily work around it just by having an `async Task Main` that awaits it[.]

Another option would be to generate code that effectively performs the workaround automatically. So this code:

```c#
static MyTask Main() { … }
```

would compiled into:

```c#
static void Main() => MainTask().GetAwaiter().GetResult();

static async Task MainTask() => await MainMyTask();

static MyTask MainMyTask() { … }
```

This way, the generated code does not rely on some unknown `GetResult()` to block, it only relies on `TaskAwaiter.GetResult()`.

> Let's stick with the name `Main` instead of `MainAsync` for now.

What is the reasoning for this? I think I don't like it, since it trains people that they shouldn't follow the TAP naming convention.

**Comment by jnm2** [comment]
> since it trains people that they shouldn't follow the TAP naming convention.

I have somewhat strong opinions on interfaces where *every* method name ends in `Async`. 😈
That convention was only ever because we don't have return value overloading and you might need to have an async and non-async method side by side. However, you'll never have two `Main` methods.

**Comment by paulomorgado** [comment]
# Implementing inaccessible interface members

That might be useful in designing interfaces where there's a public and a protected contract.

Today we need to use abstract classes with protected abstract members for that.
